sources/Timer.move

```diff
@@ -0,0 +1,188 @@
+address SFC {
+    module TimestampTimer {
+        use StarcoinFramework::Errors;
+        use StarcoinFramework::Signer;
+            use StarcoinFramework::Timestamp;
+
+        const E_INITIALIZED: u64 = 0;
+        const E_NOT_INITIALIZED: u64 = 1;
+
+        const MAX_U64: u64 = 18446744073709551615u64;
+
+        /// An timestamp based on wall clock timestamp;
+        struct Timer<phantom T> has key {
+            deadline: u64,
+        }
+
+        public fun init<T>(account: &signer) {
+            assert!(!exists<Timer<T>>(Signer::address_of(account)), Errors::already_published(E_INITIALIZED));
+            move_to(account, Timer<T>{ deadline: 0 });
+        }
+
+        spec init {
+            aborts_if exists<Timer<T>>(Signer::address_of(account));
+            ensures exists<Timer<T>>(Signer::address_of(account));
+        }
+
+        public fun set_deadline<T>(owner: address, deadline: u64) acquires Timer {
```

Comment:
I think all the public functions  except `init` need a witness param to make sure they are only called in the caller's module. Just like:
`public fun set_deadline<T>(owner: address, deadline: u64, _witness: &T) acquires Timer`.

Is this right? @jolestar

Reply:
This confuse me at the beginning, and after do a little research, I understood what you mean.

Refer to the documentation of Move: [[Privileged Struct Operations](https://move-language.github.io/move/structs-and-resources.html#privileged-struct-operations)](https://move-language.github.io/move/structs-and-resources.html#privileged-struct-operations).

> - Struct types can only be created ("packed"), destroyed ("unpacked") inside the module that defines the struct.
> - The fields of a struct are only accessible inside the module that defines the struct.

I think it's necessary to add `_witness` to functions that with `borrow_global_mut` to prevent someone else change the timer, but I'm not sure do we need prevent some else to access the timer.

Reply:
1. 
```
public fun set_deadline<T>(owner: address, deadline: u64) 
```
Allow anyone can change anyone's timer of any Type.

2. 
```
public fun set_deadline<T>(owner: address, deadline: u64, _witness: &T) 
```

Allow the module which defines T can change anyone's timer of type T.

3.
```
public fun set_deadline<T>(owner: &signer, deadline: u64) 
```

Allow the owner can change the owner's timer of any Type.

4. 
```
public fun set_deadline<T>(owner: &signer, deadline: u64, _witness: &T) 
```

Allow the module which defines T can allow the owner to change the owner's timer of type T.


I think option 2 should be what we expected, and all writing functions, `set_deadline`, and `reset`, should need a witness param, including the `init` function. 

Because the T module maybe need to do more init before Timer init.